ENH: enable OpenBLAS on windows.

xoviat opened the pull request:
Previously, #9431 allowed distutils to make use of gfortran to create fortran wrappers that worked with MSVC. For a quick review:
- Distutils detects objects that cannot be linked with an MSVC lib.
- Distutils compiles unlinkable objects into a DLL
- Distutils links Python extension against the DLL
- Distutils writes `__config__` that includes the DLLs in the PATH.

Here, we use distutils the link against the "unlinkable object" openblas.a. Distutils compiles openblas with gfortran as a DLL and then links against it. As a result of using OpenBLAS, performance restrictions on windows are addressed.

Because gfortran enables BLAS and fortran functionality, some new tests are run that were previously skipped. Some of these newly run tests fail. Such tests are marked as "known failures" on windows because they were previously untested on the CI, thus they are not regressions.

xoviat committed:
TST: appveyor: Enable OpenBLAS via MinGW Gfortran

xoviat commented:
@chrris Is pytest supported?

rgommers commented on appveyor.yml:
```diff
@@ -1,30 +1,183 @@
+# https://github.com/ogrisel/python-appveyor-demo/blob/master/appveyor.yml
+clone_depth: 50
 
-os: Visual Studio 2015
```
We still want to test MSVC as well, it needs to stay working. So instead of removing it all, I suggest to use a build matrix with a mix of the two build configs (without increasing the total number of builds).

xoviat replied:
Which MSVC are you concerned about? This configuration should test 2008, 2010, and 2015. The "os" has nothing to do with that.

matthew-brett replied:
Ralf - I'm guessing you mean a lapack-lite build, without OpenBLAS?

rgommers replied:
No, I didn't mean lapack-lite, that's okay to only test on Linux I think. The ``install`` and ``build_script`` part here explicitly add ``MINGW`` to the path and in the whole file there are only references to Cygwin and MinGW. So if MSVC is still being used, it's very much non-obvious (plus it's unclear what the purpose of the MinGW stuff is). 

So either something is wrong (like MSVC being dropped) or this needs a lot of comments explaining what is going on.

xoviat replied:
MinGW is used to compile the OpenBLAS library. MSVC is used to compile everything else.

xoviat replied:
If you look at the CI, MinGW is invoked once during configuration (system_info) and once to compile OpenBLAS. Everything else is `cl.exe`.

rgommers replied:
That's fine, but I don't want to have to inspect the actual CI runs to understand this config file. So I'll repeat my request for some comments (in particular inside the ``ps`` and ``build_script`` sections).

pv replied:
IIUC, all the compilers are installed in the appveyor VM regardless, and Python generally finds MSVC based on registry entries rather than PATH.

matthew-brett replied:
That's my understanding too.

rgommers replied:
@pv @matthew-brett not sure if you're arguing against including more documentation or not. My comment is not particular to the line ``os: Visual Studio 2015``. There's a lot of opaque stuff in this PR that simply needs explaining.

pv replied:
No, I'm just explaining why it works.

matthew-brett replied:
Yes, more documentation would be good.

I hate to be annoying, but I have previously done the same work in the windows-wheel-builder repo - here's a PR : https://github.com/numpy/windows-wheel-builder/pull/3 in case it's helpful.

xoviat commented:
So we have ` File "C:\Python36\lib\site-packages\numpy\f2py\tests\util.py", line 149, in build_module` responding "DLL load failed," probably due to skipping the configuration.

xoviat commented:
Would it be okay to xfail the three failing f2py tests since there is no way that they could have been run before?

charris commented:
@xoviat We don't yet support pytest.

xoviat commented:
@charris Let me clarify the question. There are additional tests run on this pull request because numpy.distutils is more capable now after my previous gfortran PR with gfortran on the PATH. 

Previously these tests were not run:
```
numpy.f2py.tests.test_block_docstring.TestBlockDocString.test_block_docstring ... SKIP: No C compiler available
numpy.f2py.tests.test_callback.TestF77Callback.test_string_callback ... SKIP: No C compiler available
numpy.f2py.tests.test_common.TestCommonBlock.test_common_block ... SKIP: No C compiler available
```

Now the *are* run but they fail, hence the red cross. So would it be acceptable to simply xfail these for now on windows?

charris commented:
You can skip tests by adding `dec` to the imports from `numpy.testing` and use `@dec.skipif(...)`.  Whether you *should* skip the tests is the question. Is something missing that needs to be added?

xoviat commented:
The error that I receive is:

```
ImportError: DLL load failed: The specified module could not be found.
```

The tests apparently invoke the fortran compiler which generates a DLL that must be on the path for the generated module to work correctly. Normally this isn't a problem because the `__config__` module automatically extends the path, but I suspect that the `__config__` module generated here is thrown away.

And then you have a couple Python 2.7 failures, because well, it's Python 2.7.

xoviat commented:
Hrm...the tests that I marked should be skipped are still run.

pv commented:
Nose runs setup() regardless of whether the test is marked skipped or not.
You'd probably need to add new setup() methods that raise SkipTest if to be
skipped, and otherwise call super().setup.

4.9.2017 6.29 "xoviat" <notifications@github.com> kirjoitti:

> Hrm...the tests that I marked should be skipped are still run.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/9645#issuecomment-326863615>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AACI5gZ8Pw8xbYBcuMSb7gkMQOd89B-Jks5se3ypgaJpZM4PLANo>
> .
>

pv commented:
Make that KnownFailure actually -- this stuff is supposed to work also on
windows.

The issue probably is that due to the extra-dll stuff, plain modules
without __config__.py just don't work with the msvc+gfortran hack.

4.9.2017 6.29 "xoviat" <notifications@github.com> kirjoitti:

> Hrm...the tests that I marked should be skipped are still run.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/9645#issuecomment-326863615>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AACI5gZ8Pw8xbYBcuMSb7gkMQOd89B-Jks5se3ypgaJpZM4PLANo>
> .
>

xoviat committed:
TST: fix failures:

1. fail tests related to DLL load failure as they were previously untested.
2. fix have_compiler to return false on old compilers
3. xfail some tests that were not working on old Python versions.

xoviat commented:
I've updated the description; the PR should be ready for review.

pv reviewed (commented):
The overall build config is the same as for Scipy, so I think everything is fine from implementation POV.

There are some outdated comments, and the non-f2py xfails could use extra comment.

xoviat committed:
appveyor: update documentation

xoviat commented on appveyor.yml:
```diff
@@ -1,30 +1,181 @@
+      PYTHON_ARCH: 64
+      TEST_MODE: full
+
+    - PYTHON: C:\Python27
```
I don't think they're needed but they could be used to compare differences between the released wheels and the wheels used for testing. A major advantage of this appveyor configuration is that you can actually download the built wheels and inspect them yourself.

xoviat commented on numpy/ma/tests/test_core.py:
```diff
@@ -3605,6 +3607,8 @@ def test_varstd(self):
                                 mX[:, k].compressed().std())
 
+    @dec.knownfailureif(sys.platform=='win32' and sys.version_info < (3, 6),
+                        msg='Fails on Python < 3.6')
```
No, but this appears to have not been tested on the CI before. It's probably a new discovery.

xoviat replied:
So I opened a new issue for this and updated the description with the issue number.

pv replied:
Thanks. I now remember these also appeared on Scipy, so it's probably some issue in old MSVC CRTs.

pv commented on numpy/testing/tests/test_utils.py:
```diff
@@ -263,6 +263,8 @@ def test_complex(self):
         self._test_not_equal(x, y)
 
+    @dec.knownfailureif(sys.version_info < (3, 0),
+                        msg="Error message different on Python 2")
```
Or maybe it's only on win32 (in which case there should be a platform check in the knownfailureif).

pv replied:
From log: https://ci.appveyor.com/project/charris/numpy/build/1.0.6743/job/b9ro4kpqo4h29ljf
```
AssertionError: '\nArrays are not equal\n\n(shapes (2L,), (1L, 2L) mismatch)\n x: array([1, 2])\n y: [repr failed for <matrix>: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()]' != '\nArrays are not equal\n\n(shapes (2,), (1, 2) mismatch)\n x: array([1, 2])\n y: [repr failed for <matrix>: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()]'
```
Probably would be simple to fix to accept also a second variant
`msg2 = msg.replace("(1, 2)", "(1L, 2L)")`

xoviat replied:
Good catch.

xoviat replied:
@pv To clarify, should we replace the message only on Python2 win32?

pv replied:
I'd maybe accept either one, `assert_(err == msg or err == msg2, err)`.
iirc the Ls are due to the integer sizes on win32-32 so they could also appear on other plats.

xoviat committed:
tests: core: update failure message

xoviat commented on numpy/distutils/tests/test_system_info.py:
```diff
@@ -68,7 +68,7 @@ def have_compiler():
             if not compiler.initialized:
                 compiler.initialize()  # MSVC is different
-        except DistutilsError:
+        except (DistutilsError, ValueError):
```
BTW this is because MSVC 2008 is a crappy compiler.

xoviat commented on numpy/f2py/tests/util.py:
```diff
@@ -319,6 +319,7 @@ class F2PyTest(object):
     module = None
     module_name = None
 
+    @dec.knownfailureif(sys.platform=='win32', msg='Fails with MinGW64 Gfortran')
```
Where would the best place be to put the comments? There are multiple tests all with the same cause.

pv replied:
Maybe open an issue and just put the issue number.

xoviat committed:
TST: utils: fix test_error_msg

Allow the error message to contain "large" arrays.

pv commented on numpy/testing/tests/test_utils.py:
```diff
@@ -263,21 +263,22 @@ def test_complex(self):
-                str(e),
-                "\nArrays are not equal\n\n"
+            msg = str(e)
+            msg2 = msg.replace("shapes (2,), (1, 2)", "shapes (2L,), (1L, 2L)")
```
Arguments the other way around.
Also, the try: except: isn't necessary if you do replacement in str(e).

xoviat replied:
I was not thinking here; I copied the replace statement directly from the comment without looking at it myself.

xoviat committed:
TST: util: update documentation

xoviat committed:
TST: common: update documentation

xoviat committed:
TST: block docstring: update comment

xoviat committed:
TST: callback: update comment

xoviat committed:
TST: util: fix test_error_message

use the correct replace() order

xoviat committed:
:art:

xoviat commented:
@rgommers Do you have any other questions about this?

pv commented on numpy/testing/tests/test_utils.py:
```diff
@@ -267,15 +267,18 @@ def test_error_message(self):
+            try:
+                self.assertEqual(msg, msg_reference)
+            except AssertionError:
+                self.assertEqual(msg2, msg_reference)
```
The try: except is not needed. You make it simpler via `msg = msg.replace(...)`

xoviat committed:
TST: utils: fix string continuation

rgommers commented on appveyor.yml:
```diff
@@ -1,30 +1,182 @@
+      }
+
+install:
+  - C:\cygwin\bin\du -hs "%LOCALAPPDATA%\pip\Cache"
```
Need a comment above this line, I think something like 

    This `install` step only builds and installs OpenBLAS plus build/test dependencies.
    SciPy itself is built with `build_script`. Cygwin is only needed for WHAT?

pv replied:
Cygwin is probably only needed for the unix tools (du, find etc) used for cleaning up the pip cache.

xoviat replied:
Cygwin has nothing to do with the build. This comes from experience dealing with appveyor caches.

rgommers commented on appveyor.yml:
```diff
@@ -1,30 +1,182 @@
+  - C:\cygwin\bin\du -hs "%LOCALAPPDATA%\pip\Cache"
+
+on_finish:
+  - ps: |
```
Add comment, uploads what to where? Not needed for all Appveyor scripts, so good to know why here

rgommers commented on appveyor.yml:
```diff
@@ -1,30 +1,182 @@
+      mkdir dist
+      pip wheel -v -v -v --wheel-dir=dist .
+
+      ls dist -r | Foreach-Object {
```
would be good to have a comment here, not sure what ``PushArtifact`` does and why there's a ``Foreach-Object`` (there should be only one wheel I think, built with ``pip wheel -v -v -v --wheel-dir=dist .`` above?)

rgommers commented on appveyor.yml:
```diff
@@ -1,30 +1,182 @@
+      } Else {
+          $OPENBLAS = $env:OPENBLAS_64
+      }
+      $clnt = new-object System.Net.WebClient
```
A comment here or above ``ps`` would be useful. I think it's "downloads an already built OpenBLAS library from the path given by ``$OPENBLAS``". That build was done from repo XXX.

xoviat replied:
I actually don't know the repo but matthew in in charge of these builds.

xoviat replied:
It's here: https://github.com/matthew-brett/build-openblas

xoviat replied:
I updated the comment above the `ps` block.

rgommers reviewed (changes requested):
Thanks @xoviat, clearer already. I have put some comments for all places that are still hard to understand.

xoviat committed:
appveyor: update documentation

rgommers commented:
Thanks @xoviat, all LGTM now

matthew-brett commented:
I think this one's ready to merge?

charris commented:
Let's give it a shot. The folks who know more about this than me have signed off.

charris commented:
Thanks @xoviat .
